### ISSUE TITLE
Should match on bool with if-s on match arms be considered idiomatic?

### ISSUE THREAD

**Issue by eddyp** [rust-lang/rust-clippy#5753]
I tried this code:

```rust
let (is_straight, opt_straight_rank) = match are_consecutive(&cards[1..]) {
    true if cards[0].rank - 1 == cards[1].rank => (true, Some(cards[0].rank)),
    true if cards[0].rank == 14 /* Ace */ && cards[1].rank == 5 => (true, Some(cards[1].rank)),
    _ => (false, None),
};
```

I expected no issues about the match on `bool` since the `true` match arms are conditioned and have some different content.

Instead, i got this error, which I think is suggesting to turn the match in a less readable pair of if-s (note the opt_straight_rank is bound to a different cards' `.rank`:

```
warning: you seem to be trying to match on a boolean expression
   --> src/lib.rs:196:48
    |
196 |           let (is_straight, opt_straight_rank) = match are_consecutive(&cards[1..]) {
    |  ________________________________________________^
197 | |             true if cards[0].rank - 1 == cards[1].rank => (true, Some(cards[0].rank)),
198 | |             true if cards[0].rank == 14 /* Ace */ && cards[1].rank == 5 => {
199 | |                 (true, Some(cards[1].rank))
200 | |             }
201 | |             _ => (false, None),
202 | |         };
    | |_________^
    |
    = note: `#[warn(clippy::match_bool)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#match_bool
```

I'm unsure if this is indeed a bug, but I found no option of an issue of an "inquiry" flavor.

I feel like the lint should be considering the if conditions on each arm are different/unrelated and/or if there are more than 2 arms of the match

Am I wrong? Is the match writable in a more compact/clear way without matching on bool? I like the current form as it makes it clear how the `is_straight` is working for both the regular condition as well as for the exception case of the straight starting with a 5 and ending with an ace.

**Comment by flip1995** [comment]
I think we should indeed allow it if one of the match arms has a condition.

**Comment by eddyp** [comment]
I think the condition to allow can be even more complex because:

1. if the match has just 2 arms, and the non-default one has an extra if you could write an equivalent `if`, which I feel should be recommended
1. if the match has more than 2 arms, most likely the conditions on the arm `if` are not be connected to the same logical thing, as is the case in my example; still, when there are more than 2 arms, the equivalent `if` code is likely more verbose.

**Comment by camsteffen** [comment]
I think your example is elegant in a way but at the same time it feels wrong to match on a boolean and I don't think it should be idiomatic.

Let me add another example:

```rust
match () {
    () if some_condition => foo,
    () if other_condition => bar,
}
```

This example feels wrong because the `match` is unnecessary. It should be rewritten with if/else.

I think that your example should not be considered idiomatic for the same reason. The `match` is not necessary and you could easily rewrite it with if/else.

In other words, `match` is not being used for its primary feature, but only being used for its secondary feature which, by itself, is equivalent to if/else. And in that case, if/else should be preferred.

The counterpoint is that using `match` avoids one level of nesting and a couple lines of code. But I still think if/else is easier to read cognitively.

**Comment by alhubanov** [comment]
I agree with @eddyp that the lint should not apply to match expressions with more than two arms. I tried to propose a change to this effect. 

If it is deemed that the lint should, indeed, apply regardless of how many arms there are, then perhaps this issue should be closed?
